federatedrec/factorization_machine/hetero_factorization_machine/hetero_fm_base.py

```diff
@@ -0,0 +1,114 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+#
+#  Copyright 2019 The FATE Authors. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import numpy as np
+
+from federatedml.util import consts
+from federatedml.secureprotol import PaillierEncrypt
+from federatedml.protobuf.generated import fm_model_meta_pb2
+from federatedrec.param.factorization_machine_param import HeteroFactorizationParam
+from federatedrec.factorization_machine.base_factorization_machine import BaseFactorizationMachine
+from federatedrec.transfer_variable.transfer_class.hetero_fm_transfer_variable import HeteroFMTransferVariable
+
+
+class HeteroFMBase(BaseFactorizationMachine):
+    def __init__(self):
+        super().__init__()
+        self.model_name = 'HeteroFactorizationMachine'
+        self.model_param_name = 'HeteroFactorizationMachineParam'
+        self.model_meta_name = 'HeteroFactorizationMachineMeta'
+        self.mode = consts.HETERO
+        self.aggregator = None
+        self.cipher = None
+        self.batch_generator = None
+        self.gradient_loss_operator = None
+        self.converge_procedure = None
+        self.model_param = HeteroFactorizationParam()
+
+    def _init_model(self, params):
+        super()._init_model(params)
+        self.encrypted_mode_calculator_param = params.encrypted_mode_calculator_param
+        self.cipher_operator = PaillierEncrypt()
+        self.transfer_variable = HeteroFMTransferVariable()
+        self.cipher.register_paillier_cipher(self.transfer_variable)
+        self.converge_procedure.register_convergence(self.transfer_variable)
+        self.batch_generator.register_batch_generator(self.transfer_variable)
+        self.gradient_loss_operator.register_gradient_procedure(self.transfer_variable)
+
+    def transform(self, data_inst):
```

Comment:
Does this function really need here ? In HeteroLR, this function is called by dnn_lr, but it seems like that federated recommendation will not be invoked that?

Reply:
Removed transform. Thanks.